scripts/test/Muon/loadfile_presenter_test.py

```diff
@@ -0,0 +1,510 @@
+
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=self.load_failure)
+        self.view.set_file_edit("C:\dir2\EMU000123.nxs")
+
+        self.presenter.handle_file_changed_by_user()
+        self.Runner(self.presenter._load_thread)
+
+        self.assertEqual(self.model.loaded_filenames, ["C:/dir1/EMU0001234.nxs"])
+        self.assertEqual(self.model.loaded_workspaces, [[1]])
+        self.assertEqual(self.model.loaded_runs, [1234])
+
+        self.assertEqual(self.view.reset_edit_to_cached_value.call_count, 1)
+
+
+class LoadFileWidgetPresenterMultipleFileModeTest(unittest.TestCase):
+    def run_test_with_and_without_threading(f):
+
+        def run_twice(self):
+            f(self)
+            self.setUp()
+            self.presenter._use_threading = False
+            f(self)
+
+        return run_twice
+
+    class Runner:
+
+        def __init__(self, thread):
+            if thread:
+                self._thread = thread
+                self._thread.finished.connect(self.finished)
+                if self._thread.isRunning():
+                    QT_APP.exec_()
+
+        def finished(self):
+            QT_APP.processEvents()
+            QT_APP.exit(0)
+
+    def setUp(self):
+        self.data = MuonLoadData()
+        self.view = BrowseFileWidgetView()
+        self.model = BrowseFileWidgetModel(self.data)
+
+        self.view.disable_load_buttons = mock.Mock()
+        self.view.enable_load_buttons = mock.Mock()
+        self.view.warning_popup = mock.Mock()
+
+        self.presenter = BrowseFileWidgetPresenter(self.view, self.model)
+        self.presenter.enable_multiple_files(True)
+
+    def mock_loading_multiple_files_from_browse(self, runs, workspaces, filenames):
+        self.view.show_file_browser_and_return_selection = mock.Mock(return_value=filenames)
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=zip(workspaces, runs))
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Multiple runs can be selected via browse and entered explicitly using the ";" separator
+    # ------------------------------------------------------------------------------------------------------------------
+
+    @run_test_with_and_without_threading
+    def test_that_cannot_load_same_file_twice_from_same_browse_even_if_filepaths_are_different(self):
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=zip([[1], [2]], [1234, 1235]))
+        self.view.show_file_browser_and_return_selection = mock.Mock(
+            return_value=["C:/dir1/file1.nxs", "C:/dir2/file1.nxs", "C:/dir2/file2.nxs"])
+
+        self.presenter.on_browse_button_clicked()
+        self.Runner(self.presenter._load_thread)
+
+        six.assertCountEqual(self, self.model.loaded_filenames, ["C:/dir1/file1.nxs", "C:/dir2/file2.nxs"])
```

Comment:
I think we should check the model outputs in a test for the model. I think here we just want to check the correct stuff is called. This would split the tests up and make them a bit easier to read

Reply:
What do you think generally about the practice of having multiple test classes in one file? Most of my tests do this to categorise the tests better and improve readability; so do we make it a rule to split them all up into separate files, or take it on a case-by-case basic?

Reply:
I think it should be case by case. The reason I said it was that the file was getting rather long and it had a nice distinction from the previous tests (so we could sensibly split it up)

Reply:
my comment here was more we can test the presenter without caring what the model does and then test that the model works. So they have individual tests and they are not tied down to the presenter (so we can use the same model in different presenters/views)